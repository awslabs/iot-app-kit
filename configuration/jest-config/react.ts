export default {
  setupFilesAfterEnv: ['jest-extended/all'],
  maxWorkers: '50%',
  moduleFileExtensions: ['js', 'ts', 'jsx', 'tsx'],
  collectCoverage: true,
  testEnvironment: 'jsdom',
  transform: {
    '.*\\.(tsx?|jsx?)$': [
      '@swc/jest',
      {
        jsc: {
          transform: {
            react: {
              runtime: 'automatic',
            },
          },
        },
      },
    ],
  },
  transformIgnorePatterns: [],
  moduleNameMapper: {
    '\\.(css|scss)$': 'identity-obj-proxy',
  },
};
