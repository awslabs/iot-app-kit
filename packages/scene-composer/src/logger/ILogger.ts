interface ILogger {
  extend(namespace: string): ILogger;
  warn(message: string, ...args: any[]): Promise<void>;
  error(message: string, error?: any, ...args: any[]): Promise<void>;
  fatal(message: string, error?: any, ...args: any[]): Promise<void>;
  info(message: string, ...args: any[]): Promise<void>;
  verbose(message: string, ...args: any[]): Promise<void>;
}

export default ILogger;
