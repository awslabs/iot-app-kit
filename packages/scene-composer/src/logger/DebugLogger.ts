import debug from 'debug';

import ILogger from './ILogger';

class DebugLogger implements ILogger {
  private log: debug.Debugger;

  constructor(log: debug.Debugger | string) {
    if (typeof log === 'string') {
      this.log = debug(log);
    } else {
      this.log = log;
    }
  }

  extend(namespace: string): ILogger {
    return new DebugLogger(this.log.extend(namespace));
  }

  async warn(message: string, ...args: any): Promise<void> {
    return this.write('WARN', message, ...args);
  }

  async error(message: string, error?: Error, ...args: any): Promise<void> {
    return this.write('ERROR', message, error, ...args);
  }

  async fatal(message: string, error?: Error, ...args: any): Promise<void> {
    return this.write('FATAL', message, error, ...args);
  }

  async info(message: string, ...args: any): Promise<void> {
    return this.write('INFO', message, args);
  }

  async verbose(message: string, ...args: any): Promise<void> {
    return this.write('VERBOSE', message, ...args);
  }

  private async write(level: string, message: string, ...args: any): Promise<void> {
    const logger = this.log.extend(`[${level.toUpperCase()}]`);
    return Promise.resolve(logger(message, ...args));
  }
}

export default DebugLogger;
