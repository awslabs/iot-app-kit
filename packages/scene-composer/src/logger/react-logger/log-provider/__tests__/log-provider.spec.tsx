import React, { useContext } from 'react';
import { render } from '@testing-library/react';

import DebugLogger from '../../../DebugLogger';
import ILogger from '../../../ILogger';
import LogProvider from '..';
import LoggingContext from '../../contexts/logging';

jest.mock('../../../DebugLogger');

const MockLogger = {
  extend() {
    return MockLogger;
  },
  warn: jest.fn(),
  error: jest.fn(),
  fatal: jest.fn(),
  info: jest.fn(),
  verbose: jest.fn(),
};

describe('<LogProvider />', () => {
  it('Log unhandled errors in child components', () => {
    const Child = () => {
      throw new Error('expected error');
    };

    const { container } = render(
      <LogProvider namespace='test-test' ErrorView={() => <div>Something went wrong</div>}>
        <Child />
      </LogProvider>,
    );

    expect(container.outerHTML).toMatchSnapshot();
  });

  it('should allow overriding the default logger', () => {
    const Child = () => {
      throw new Error('expected error');
    };

    const logger = MockLogger as ILogger;

    const { container } = render(
      <LogProvider namespace='test-test' logger={logger} ErrorView={() => <div>Something went wrong</div>}>
        <Child />
      </LogProvider>,
    );

    expect(container.outerHTML).toMatchSnapshot();
  });

  it('should support nesting loggers', () => {
    const log = jest.fn();
    const top = 'This is a log';
    const inner = 'this is another log';

    (DebugLogger as any).mockImplementation(() => ({
      info: log,
      verbose: jest.fn(),
      extend: (namespace: string) => new DebugLogger(namespace),
    }));

    const LoggingComponent = ({ message }: { message: string }) => {
      const { log } = useContext(LoggingContext);

      log?.info(message);

      return <div>Logged</div>;
    };

    render(
      <LogProvider namespace='Foo'>
        <LoggingComponent message={top} />
        <LogProvider namespace='Bar'>
          <LoggingComponent message={inner} />
        </LogProvider>
      </LogProvider>,
    );

    expect(log).toBeCalledWith(top);
    expect(log).toBeCalledWith(inner);
  });
});
