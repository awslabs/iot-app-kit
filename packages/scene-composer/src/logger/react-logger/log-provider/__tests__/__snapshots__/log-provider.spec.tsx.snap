// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LogProvider /> Log unhandled errors in child components 1`] = `"<div><div>Something went wrong</div></div>"`;

exports[`<LogProvider /> should allow overriding the default logger 1`] = `"<div><div>Something went wrong</div></div>"`;
